File: cpp/sa/TSP-simulated-annealing.cpp
//TSP Simulate Annealing implementation
//Author: David Talamas
//CS345 assignment 6
//May 10, 2016

#include <iostream>
#include <vector>
#include <fstream>
#include <string>
#include <stdlib.h>
#include <math.h>
#include <time.h>

using namespace std;

//Graph represented as a vector of points (x,y)
class Graph{
public: 
  vector<pair<int, int> > nodes;
};

//Get euclidian distance between two points
double dist(pair<int, int> a, pair<int, int> b){
  double xd = a.first - b.first;
  double yd = a.second - b.second;
  return sqrt(xd * xd + yd * yd);
}

double pathCost(vector<pair<int, int> > &nodes, vector<int> &path){
  double cost = 0;
  for (int i = 0; i < nodes.size() - 2; i++){
    cost += dist(nodes[path[i]], nodes[path[i+1]]);
  }
  cost += dist(nodes[path[nodes.size()-2]], nodes[path[0]]);
  return cost;    
}

//Prints tour
void printTour(vector<int> &tour){
  cout<<"Tour:\n";
  for (vector<int>::iterator it = tour.begin(); it != tour.end(); ++it){
    cout<<*it<<"\n";
  }
  return;
}

//Gets randomly a neighboring state
vector<int> nextState(vector<int> s){
  vector<int> newState = s;
  int a,b;
  a = rand() % s.size();
  b = rand() % s.size();
  int temp = newState[a];
  newState[a] = newState[b];
  newState[b] = temp;
  return newState;
}

//returns true with probability P = exp(-(cost2-cost1)/T);
bool flipCoin(double cost1, double cost2, double T){
  double p = exp(-(cost2-cost1)/T);
  double u = ((double) rand() / (RAND_MAX));
  if (u < p){
    return true;
  }
  else {
    return false;
  }
}

int main(int argc, char* argv[]){
  srand(time(NULL));
  if (argc != 2){
    cout<<"Wrong args";
    return 0;
  }
  //Get the data from the given file
  char *path = argv[1];
  ifstream f; 
  f.open(path, ios::in);
  Graph g;
  int size = 0;
  if (f.is_open()){
    string line;
    while (getline(f, line)){
      if (line.length()> 9 && line.substr(0,9).compare("DIMENSION")==0){
	size = atoi(line.substr(11,line.length()-11).c_str());
	break;
      }
    }
    g.nodes.resize(size + 1);
    while (getline(f, line)){
      if (line.length()>3 && line[2]<= '9' && line[2]>= '0'){
	int index = atoi(line.substr(0,3).c_str());
	int x = atoi(line.substr(4,3).c_str());
	int y = atoi(line.substr(8,3).c_str());
	g.nodes[index].first = x;
	g.nodes[index].second = y;
      }
    }
  }
  f.close();

  //Choose initial tour randomly
  vector<int> tour(size, 0);
  int nums[280] = {0};
  for (int i = 1; i <= size; i++){    
    //tour[i-1] = i;
    while(1){
      int r = rand() % size;
      if (nums[r] == 0){
	tour[i-1] = r;
	nums[r] = 1;
	break;
      }
    }
  }

  
  double min = pathCost(g.nodes, tour);
  vector<int> minTour = tour;
  double T = 100; 
  
  //Simulated annealing
  while (T > .0001){
    //Iterate 100 times for each temperature
    for (int q = 0; q <100; q++){
      vector<int> s_ = nextState(tour);
      double prevCost = pathCost(g.nodes, tour);
      double newCost = pathCost(g.nodes, s_);
      if (newCost < prevCost){
	tour = s_;
	if (newCost < min){
	  min = newCost;
	  minTour = tour;
	}
      } 
      else if (flipCoin(prevCost, newCost, T)) {
	tour = s_;
      }
    }
    T *= .9999;
  }

  //Print results
  cout<<"Cost: "<<pathCost(g.nodes, minTour)<<" \n";
  printTour(tour);
  return 0;
  

}

File: python/sa/TSP-simulated-annealing.py
###############################################################################
# Mike Feilbach
# University of Texas at Austin
# CS 345, Spring 2016
# Prof. Rellermeyer
###############################################################################

import sys
import math
import random
import copy


###############################################################################
# Each Node represents a pyhsical location in the TSP problem.
###############################################################################
class Node:

	def __init__(self, name, x_coord, y_coord):
		self.name = name
		self.x_coord = x_coord
		self.y_coord = y_coord
   
	def getName(self):
   		return self.name

	def getXCoordinate(self):
		return self.x_coord

	def getYCoordinate(self):
		return self.y_coord

	def __repr__(self):
		return '[' + self.name + ', ' + str(self.x_coord) + ', ' + str(self.y_coord) + ']\n'


###############################################################################
# Returns the Edclidean distance between two Nodes.
###############################################################################
def distanceBetween(NodeA, NodeB):
	
	return math.sqrt(((NodeB.getXCoordinate() - NodeA.getXCoordinate()) ** 2) 
		+  ((NodeB.getYCoordinate() - NodeA.getYCoordinate()) ** 2))


###############################################################################
# Parses the a280.tsp file (the TSP input file). Returns a list of Nodes,
# where each Node represents a location in the file.
###############################################################################
def parseTSPFile(TSPfileName):
	
	file = open(TSPfileName, 'r')

	# List of Nodes to return.
	listOfNodes = []

	# Parse line by line.
	for line in file:
		splitList = line.split();

		# If we received three digits, it is a location in the TSP file.
		if all(item.isdigit() for item in splitList):
			if len(splitList) == 3:

				# Make a node out of this location.
				listOfNodes.append(Node(splitList[0], float(splitList[1]), float(splitList[2])))

	return listOfNodes


###############################################################################
# Returns the distance of this tour. The tour goes from the first Node
# in the list to the second Node, to the third Node,..., to the last Node, and 
# back to the first Node.
###############################################################################
def tourDistance(listOfNodes):
	
	distance = 0;

	# Get the distance from Node 1 to Node 2,... to Node N, where N is
	# the last node in the list.
	for i in range(0, len(listOfNodes) - 1):

		distance = distance + distanceBetween(listOfNodes[i], listOfNodes[i + 1])


	# Get the distance from Node N back to Node 1.
	# Note: no error checking here, beacuse the input file is assumed to have
	# locations in it.
	distance += distanceBetween(listOfNodes[len(listOfNodes) - 1], listOfNodes[0])

	return distance


###############################################################################
# Generates a neighboring tour, given a current tour. For now, this will
# simply swap two locations on the tour. For example, [1, 2, 3, 4] might go to
# [1, 4, 3, 2].
###############################################################################
def getNeighborTour(listOfNodes):

	newList = copy.deepcopy(listOfNodes)

	# Generate two random numbers corresponding to two elements in this
	# list of Nodes (in this tour).
	random1 = random.randint(0, len(listOfNodes) - 1)
	random2 = random.randint(0, len(listOfNodes) - 1)

	newList[random1], newList[random2] = newList[random2], newList[random1]

	return newList


###############################################################################
# Exponential coin flip, as defined in the simulated annealing algorithm
# at: https://www.seas.gwu.edu/~simhaweb/champalg/tsp/tsp.html
###############################################################################
def exponentialCoinFlip(currState, neighborState, currTemperature):

	probabilityToFlip = math.exp((-1 * (tourDistance(neighborState) - tourDistance(currState))) / currTemperature)

	uniformRandom = random.random()

	if (uniformRandom < probabilityToFlip):
		return True
	else:
		return False


###############################################################################
# Given an initial tour, initial temp, number of iterations, and cooling rate,
# performed simulated annealing on the tour and returns the final solution 
# tour.
###############################################################################
def performSimulatedAnnealing(listOfNodes, startingTemp, numIterations, coolingRate):

	# Do not modify the argument.
	s = copy.deepcopy(listOfNodes)

	# Create a random initial state.
	for i in range (0, 500):
		s = getNeighborTour(s)

	# This is the initial cost.
	#print "INITIAL COST: ", tourDistance(s)

	# The best tour yet is the initial one to start.
	# Must deep copy to kill the reference.
	minTour = copy.deepcopy(s)
	minCost = tourDistance(s)

	temperature = startingTemp

	for i in range (0, numIterations):
		
		# Get a random neighboring tour.
		# Note: this does not modify s and returns a new list.
		newS = getNeighborTour(s)

		if (tourDistance(newS) < tourDistance(s)):

			# The neighboring tour is better. Jump to it.
			s = copy.deepcopy(newS)

			# See if it is the best so far.
			if (tourDistance(newS) < minCost):
				minCost = tourDistance(newS)
				minTour = copy.deepcopy(newS)

				#print "Updated min tour: old dist: ", minCost, "new dist: ", tourDistance(newS), "iterations left: ", (numIterations - i)
		
		elif exponentialCoinFlip(s, newS, temperature):

			# Jump to the neighbor, even though it's worse.
			s = copy.deepcopy(newS)

		# Decrease temperature.
		temperature = temperature * coolingRate

	return minTour


def main():

	# Check for TSP input file.
	if len(sys.argv) < 2:
			print "No input file given."
			sys.exit(1)

	# Parse the TSP input file. Get a list of Nodes, where each Node
	# represents a location in the TSP problem within the file.
	#print "Parsing file: ", sys.argv[1]
	listOfNodes = parseTSPFile(sys.argv[1])


	#########################################################
	# Example experiment with:
	#      initial temperature:  200,000
	#      number of iterations: 40,000
	#      cooling rate:         .999
	#########################################################

	# Get a solution to TSP via the simulated annealing algorithm.
	solutionList = performSimulatedAnnealing(listOfNodes, 200000, 40000, .999)
	print 'Solution tour distance: ', tourDistance(solutionList)


	########################################
	# Experimental data below.
	########################################

	# 8359
	#solutionList = performSimulatedAnnealing(listOfNodes, 200000, 40000, .99)
	#print tourDistance(solutionList)

	# 8720
	#solutionList = performSimulatedAnnealing(listOfNodes, 100000, 40000, .999)
	#print tourDistance(solutionList)

	# 7843
	#solutionList = performSimulatedAnnealing(listOfNodes, 100000, 40000, .99)
	#print tourDistance(solutionList)

	# 8458
	#solutionList = performSimulatedAnnealing(listOfNodes, 10000, 40000, .99)
	#print tourDistance(solutionList)

	# 8326
	#solutionList = performSimulatedAnnealing(listOfNodes, 50000, 40000, .99)
	#print tourDistance(solutionList)

	# 8897
	#solutionList = performSimulatedAnnealing(listOfNodes, 5000, 40000, .99)
	#print tourDistance(solutionList)

	# 8456
	#solutionList = performSimulatedAnnealing(listOfNodes, 10000, 60000, .99)
	#print tourDistance(solutionList)

	# Not good.
	#solutionList = performSimulatedAnnealing(listOfNodes, 200000, 40000, .9999)
	#print tourDistance(solutionList)

	# 7422
	#solutionList = performSimulatedAnnealing(listOfNodes, 200000, 80000, .999)
	#print tourDistance(solutionList)

	# 8638
	#solutionList = performSimulatedAnnealing(listOfNodes, 10000, 40000, .99)
	#print tourDistance(solutionList)

	# Not good.
	#solutionList = performSimulatedAnnealing(listOfNodes, 10000, 10000, .99)
	#print tourDistance(solutionList)

	# Not good.
	#solutionList = performSimulatedAnnealing(listOfNodes, 200000, 10000, .99)
	#print tourDistance(solutionList)

	# Very not good.
	#solutionList = performSimulatedAnnealing(listOfNodes, 100000, 80000, .9999)
	#print tourDistance(solutionList)


if __name__ == '__main__':
	main()





